src/mp3.c

```diff
@@ -0,0 +1,354 @@
+static	sf_count_t	mp3_read_2i		(SF_PRIVATE * psf, int * ptr, sf_count_t len) ;
+static	sf_count_t	mp3_read_2f		(SF_PRIVATE * psf, float * ptr, sf_count_t len) ;
+static	sf_count_t	mp3_read_2d		(SF_PRIVATE * psf, double * ptr, sf_count_t len) ;
+static	sf_count_t	mp3_read_seek		(SF_PRIVATE *psf, int whence, sf_count_t offset) ;
+static	ssize_t		mp3_read_sf_handle	(void * handle, void * buffer, size_t bytes) ;
+static	off_t		mp3_seek_sf_handle	(void * handle, off_t offset, int whence) ;
+static	int		mp3_format_to_encoding	(int encoding) ;
+
+static	char *		mp3_init_lame		(void) ;
+static	void		mp3_close_lame		(void) ;
+
+static	int		mp3_write_open		(SF_PRIVATE * psf) ;
+static	int		mp3_write_close		(SF_PRIVATE * psf) ;
+static	sf_count_t	mp3_write_2s		(SF_PRIVATE *psf, const short *ptr, sf_count_t len) ;
+
+// FIXME: This initialisation should have a better hook
+static int mpg123_initialised = 0 ;
+
+/*------------------------------------------------------------------------------
+** Public function.
+*/
+
+int
+mp3_open (SF_PRIVATE * psf)
+{	psf->codec_data = NULL ;
+	switch (psf->file.mode) {
+	case (SFM_READ) :
+		return mp3_read_open (psf) ;
+	case (SFM_WRITE) :
+		return mp3_write_open (psf) ;
+	default:
+		return SFE_UNIMPLEMENTED ;
+	}
+}
+
+/*------------------------------------------------------------------------------
+*/
+
+// FIXME: use mpg123 error string reporting
+static int
+mp3_read_open (SF_PRIVATE * psf)
+{
+	int decoder_err = MPG123_OK ;
+	mpg123_handle * decoder ;
+	if (mpg123_initialised == 0)
+	{	int decoder_init_err = mpg123_init () ;
+		if (decoder_init_err != MPG123_OK)
+		{	psf_log_printf (psf, "Failed to init mpg123.\n") ;
+			return SFE_UNIMPLEMENTED ; // FIXME: semantically wrong return code
+		}
+	}
+	mpg123_initialised++ ;
+
+	decoder = mpg123_new (NULL, &decoder_err) ;
+
+	if (decoder_err == MPG123_OK)
+		decoder_err = mpg123_replace_reader_handle (
+			decoder, mp3_read_sf_handle, mp3_seek_sf_handle, NULL) ;
+	psf->fileoffset = 0 ; // FIXME: Remove this once the seek fixes are in
+	if (decoder_err == MPG123_OK)
+		decoder_err = mpg123_open_handle (decoder, psf) ;
+
+	if (decoder_err == MPG123_OK)
+		decoder_err = mp3_read_header (psf, decoder) ;
+
+	if (decoder_err != MPG123_OK)
+	{	psf_log_printf (psf, "Failed to initialise mp3 decoder.\n") ;
+		return SFE_UNIMPLEMENTED ; // FIXME: semantically wrong return code
+	}
+
+	psf->dataoffset = psf_ftell (psf) ;
+	psf->datalength = psf->filelength - psf->dataoffset ;
+	psf->codec_data = decoder ;
+
+	psf->container_close = mp3_read_close ;
+	psf->seek = mp3_read_seek ;
+
+	psf->read_short = mp3_read_2s ;
+	psf->read_int = mp3_read_2i ;
+	psf->read_float = mp3_read_2f ;
+	psf->read_double = mp3_read_2d ;
+
+	return 0 ;
+}
+
+static int
+mp3_read_close (SF_PRIVATE * psf)
+{	mpg123_handle * decoder = psf->codec_data ;
+	if (decoder != NULL)
+	{	// mpg123_close(decoder); <- Not sure if we need this?
```

Comment:
If it doesn't blow up, call it anyway.

Reply:
IIRC it did, which is why it was commented out, but I'll check.